### PR TITLE
GH-123391: Code generators: Track stack and locals in `if` statements.

### PR DESCRIPTION
**Pull request by markshannon**
This PR adds little functionality, but it will enable us to do much more effective stack spilling in the future.

I've deliberately left the spilling functionality for a later PR to make this PR easier to review.

This PR:

* Adds a new `Storage` class which combines the state of the stack and the local variables within a micro-op.
* Adds copying and merging functionality to that class, and the underlying stack and locals, to support tracking state across divergent flow.
* Splits and merges the storage in `if` statements.

The resulting generated code is basically the same as we have now.

To simplify parsing, the code generator also enforces PEP 7 rules for braces.
Most of the changes in bytecodes.c are a result of this change.

* Issue: gh-123391

### CODE REVIEW

File: Tools/cases_generator/stack.py [hunk 1]
```diff
@@ -370,3 +404,58 @@ def stacks(inst: Instruction | PseudoInstruction) -> Iterator[StackEffect]:
...
+                self.stack.push(out)
+                self.outputs.pop(0)
+            else:
+                break
+        undefined = ""
+        for out in self.outputs:
+            if out.defined:
+                raise StackError(
+                    f"Locals not defined in stack order. "
+                    f"Expected  '{out.name}' is defined before '{undefined}'"
```
Comment (picnixz) [inline Tools/cases_generator/stack.py:427]:
```suggestion
                    f"Locals not defined in stack order. "
                    f"Expected {out.name!r} to be defined before {undefined!r}"
```

File: Tools/cases_generator/stack.py [hunk 2]
```diff
@@ -370,3 +404,58 @@ def stacks(inst: Instruction | PseudoInstruction) -> Iterator[StackEffect]:
...
+        while self.outputs:
+            out = self.outputs[0]
+            if out.defined:
+                self.stack.push(out)
+                self.outputs.pop(0)
+            else:
+                break
+        undefined = ""
+        for out in self.outputs:
+            if out.defined:
```
Comment (picnixz) [inline Tools/cases_generator/stack.py:424]: If the first `out.defined` holds, then we don't really have anything "before" (unless this case is not possible?)

File: Tools/cases_generator/tier2_generator.py [hunk 3]
```diff
@@ -69,85 +71,100 @@ def __init__(self, out: CWriter):
...
-        stack: Stack,
+        storage: Storage,
         inst: Instruction | None,
-    ) -> None:
+    ) -> bool:
         self.out.emit_at("if ", tkn)
-        self.emit(next(tkn_iter))
+        lparen = next(tkn_iter)
+        self.emit(lparen)
+        assert(lparen.kind == "LPAREN")
```
Comment (picnixz) [inline Tools/cases_generator/tier2_generator.py:82]:
```suggestion
        assert lparen.kind == "LPAREN"
```

File: Tools/cases_generator/tier2_generator.py [hunk 4]
```diff
@@ -69,85 +71,100 @@ def __init__(self, out: CWriter):
...
-        unused: Stack,
+        storage: Storage,
         inst: Instruction | None,
-    ) -> None:
+    ) -> bool:
         self.out.emit_at("if ", tkn)
-        self.emit(next(tkn_iter))
+        lparen = next(tkn_iter)
+        self.emit(lparen)
+        assert(lparen.kind == "LPAREN")
```
Comment (picnixz) [inline Tools/cases_generator/tier2_generator.py:118]:
```suggestion
        assert lparen.kind == "LPAREN"
```

File: Tools/cases_generator/generators_common.py [hunk 5]
```diff
@@ -77,45 +130,64 @@ def __init__(self, out: CWriter):
...
-        unused: Stack,
+        storage: Storage,
         inst: Instruction | None,
-    ) -> None:
+    ) -> bool:
         self.out.emit_at("DEOPT_IF", tkn)
-        self.out.emit(next(tkn_iter))
+        lparen = next(tkn_iter)
+        self.emit(lparen)
+        assert(lparen.kind == "LPAREN")
```
Comment (picnixz) [inline Tools/cases_generator/generators_common.py:159]:
```suggestion
        assert lparen.kind == "LPAREN"
```

File: Tools/cases_generator/generators_common.py [hunk 6]
```diff
@@ -47,22 +75,47 @@ def write_header(
...
+        else:
+            break
+    undefined = ""
+    for out in outputs:
+        if out.defined:
+            raise analysis_error(
+                f"Locals not defined in stack order. "
+                f"Expected  '{out.name}' is defined before '{undefined}'",
+                tkn
+            )
```
Comment (picnixz) [inline Tools/cases_generator/generators_common.py:115]:
```suggestion
            raise analysis_error(
                f"Locals not defined in stack order. "
                f"Expected {out.name!r} to be defined before {undefined!r}",
                tkn
            )
```

File: Tools/cases_generator/stack.py [hunk 7]
```diff
@@ -344,6 +355,29 @@ def peek_offset(self) -> str:
     def as_comment(self) -> str:
         return f"/* Variables: {[v.name for v in self.variables]}. Base offset: {self.base_offset.to_c()}. Top offset: {self.top_offset.to_c()} */"
 
+    def copy(self) -> "Stack":
+        other = Stack()
+        other.top_offset = self.top_offset.copy()
+        other.base_offset = self.base_offset.copy()
+        other.variables = [var.copy() for var in self.variables]
+        other.defined = set(self.defined)
```
Comment (picnixz) [inline Tools/cases_generator/stack.py:363]:
```suggestion
        other.defined = self.defined.copy()
```

File: Tools/cases_generator/generators_common.py [hunk 8]
```diff
@@ -77,45 +130,64 @@ def __init__(self, out: CWriter):
...
-        stack: Stack,
+        storage: Storage,
         inst: Instruction | None,
-    ) -> None:
+    ) -> bool:
         self.out.emit_at("if ", tkn)
-        self.out.emit(next(tkn_iter))
+        lparen = next(tkn_iter)
+        self.emit(lparen)
+        assert(lparen.kind == "LPAREN")
```
Comment (picnixz) [inline Tools/cases_generator/generators_common.py:183]:
```suggestion
        assert lparen.kind == "LPAREN"
```

File: Lib/test/test_generated_cases.py [hunk 9]
```diff
@@ -281,6 +281,67 @@ def test_predictions(self):
...
+            stack_pointer[0] = res;
+            stack_pointer += 1;
+            assert(WITHIN_STACK_BOUNDS());
+            DISPATCH();
+        }
+    """
+        self.run_cases_test(input, output)
+
+
+    def test_pep7_condition(self):
```
Comment (picnixz) [inline Lib/test/test_generated_cases.py:331]: Maybe test the else branch as well?

**Review by picnixz — COMMENTED** [review]
Some nitpicks.

**Comment by markshannon** [comment]
Replaced by https://github.com/python/cpython/pull/124392

@picnixz Thanks for review. I've ported your suggestion to the replacement PR.

**Pull request closed without merging**